funsor/affine.py

```diff
@@ -0,0 +1,47 @@
+from collections import OrderedDict
+
+import opt_einsum
+import torch
+
+from funsor.interpreter import gensym
+from funsor.terms import Lambda, Variable, bint
+from funsor.torch import Tensor
+
+
+def extract_affine(fn):
+    """
+    Extracts an affine representation of a funsor, which is exact for affine
+    funsors and approximate otherwise. For affine funsors this satisfies::
+
+        x = ...
+        const, coeffs = extract_affine(x)
+        y = sum(Einsum(eqn, (coeff, Variable(var, coeff.output)))
+                for var, (coeff, eqn) in coeffs.items())
+        assert_close(y, x)
+
+    :param Funsor fn: A funsor assumed to be affine (this is not checked).
+    :return: A pair ``(const, coeffs)`` where const is a funsor with no real
+        inputs and ``coeffs`` is an OrderedDict mapping input name to a
+        ``(coefficient, eqn)`` pair in einsum form.
+    :rtype: tuple
+    """
+    # Determine constant part by evaluating fn at zero.
+    real_inputs = OrderedDict((k, v) for k, v in fn.inputs.items() if v.dtype == 'real')
+    zeros = {k: Tensor(torch.zeros(v.shape)) for k, v in real_inputs.items()}
+    const = fn(**zeros)
+
+    # Determine linear coefficients by evaluating fn on basis vectors.
+    name = gensym('probe')
+    coeffs = OrderedDict()
+    for k, v in real_inputs.items():
```

Comment:
Wrap this loop in `with funsor.memoize()` to avoid repeated tensor computations?  We should also be able to use #235 for further savings once that's implemented

Reply:
I'd prefer to implement this simple version first. Most of our funsor expressions have 1 or 2 real free variables, and I'd rather get lots of tests in place before introducing extra complexity.

Reply:
I've added a note that "users may want to run this under the memoize interpretation".